### ISSUE TITLE
[Bug] Using AAD B2C in aspnetcore 3.0 behind Azure FrontDoor results in login exception (no correlation cookies in response)

### ISSUE THREAD

**Issue by ossentoo** [AzureAD/microsoft-authentication-library-for-dotnet#1609]
**Which Version of MSAL are you using ?**

MSAL 4.3.0

**Platform**
netcore 3.0

**What authentication flow has the issue?**
* Web App
    * [x] Authorization code

Other? - please describe;

**Is this a new or existing app?**
c. This is a new app or experiment
App is in development testing, but not yet prod

**Repro**
```

/// appsettings.json
{
  "Authentication": {
    "AzureAdB2C": {
      "ClientId": "00000000-0000-0000-0000-000000000000",
      "Domain": "devlocal.b2clogin.com",
      "Tenant": "devlocal.onmicrosoft.com",
      "GraphUri": "https://graph.microsoft.com/",
      "GraphRelativePath": "v1.0/users",
      "SignUpSignInPolicyId": "b2c_1_susi",
      "ResetPasswordPolicyId": "b2c_1_reset",
      "EditProfilePolicyId": "b2c_1_edit_profile",
      "RedirectUri": "https://localhost:44387/signin-oidc",
      "ClientSecret": "980234jkl2j34k2349"
    },

```

/// csharp
```

            public void Configure(string name, OpenIdConnectOptions options)
            {
                options.ClientId = AzureAdB2CSettings.ClientId;
                options.Authority = AzureAdB2CSettings.Authority;
                options.UseTokenLifetime = true;
                options.TokenValidationParameters = new TokenValidationParameters { NameClaimType = "name" };

                options.Events = new OpenIdConnectEvents
                {
                    OnRedirectToIdentityProvider = OnRedirectToIdentityProvider,
                    OnRemoteFailure = OnRemoteFailure,
                    OnAuthorizationCodeReceived = **OnAuthorizationCodeReceived**,
                    OnTokenValidated = OnSecurityTokenValidated,
                    OnMessageReceived = OnMessageReceived
                };
            }

            private async Task OnAuthorizationCodeReceived(AuthorizationCodeReceivedContext context)
            {
                string signedInUserId = context.Principal.FindFirst(ClaimTypes.NameIdentifier).Value;

                var app = ConfidentialClientApplicationBuilder.Create(AzureAdB2CSettings.ClientId)
                    .WithB2CAuthority(AzureAdB2CSettings.Authority)
                    .WithRedirectUri(AzureAdB2CSettings.RedirectUri)
                    .WithClientSecret(AzureAdB2CSettings.ClientSecret)
                    .Build();

                new MsalStaticCache(signedInUserId, context.HttpContext).EnablePersistence(app.UserTokenCache);

                try
                {
                    var accounts = await app.GetAccountsAsync();
                    var result = await app.AcquireTokenSilent(AzureAdB2CSettings.ApiScopes.Split(' '), accounts.FirstOrDefault())
                        .ExecuteAsync();
                    context.HandleCodeRedemption(result.AccessToken, result.IdToken);
                }
                catch (MsalServiceException ex)
                {
                    Logger.LogError(ex.Message);
                }
            }
```

**Expected behavior**
I've got an aspnetcore 3.0 web app that uses Azure B2C.  In dev it has been deployed to an Azure web app and works as expected with Azure B2C and MSAL.  We can use it with a custom domain name and get the result we want.  I.e.

- user clicks on the Sign In link
- they are redirected to the MS B2C site and can sign in as expected (2FA via text message is enabled)
- After they have been authenticated, they are redirected back to our site

in the request that works (either using local debugging or i deploy the website without FrontDoor to Azure), i see these cookies:

![image](https://user-images.githubusercontent.com/7569380/73589724-2a329380-44d2-11ea-82f6-d09f49a3525a.png)

**Actual behavior**
However, when the same site is deployed behind an Azure FrontDoor, the redirect back from Azure B2C doesn't contain the aspnet correlation cookies.  This results in a **Correlation Failed** error

In this scenario, i don't see the aspnetcore correlation cookies in the response.

**Possible Solution**

We logged this via a support request on Azure and were directed to an engineer for debugging and review.  

He suggested that we need to somehow instruct our app to send heads to FrontDoor to instruct it not to cache the response that comes back after authentication.  He recommended: Cache-Control: private.

The question is, as it is MSAL that is performing the signin request on our behalf, how can we send those headers as part of the request?  I'm not sure whether this is a bug report or a feature request as I can't find any documentation which describes how to do this.

Possibly there could be some options/parameters/settings on this object method?

**ConfidentialClientApplicationBuilder.Create**

**Additional context/ Logs / Screenshots**
Add any other context about the problem here, such as logs and screebshots. Logging is described at https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/logging

![image](https://user-images.githubusercontent.com/7569380/73569894-3d5a4a80-4463-11ea-97a3-89f612e99a3b.png)

**Comment by jennyf19** [comment]
@ossentoo did you disable authentication in EasyAuth?

cc: @jmprieur

**Comment by ossentoo** [comment]
Hi @jennyf19 no, i don't think I've disabled it.  Can you explain a little about how this would work, or point me to some documentation?  I haven't heard of that option.

thanks

**Comment by jennyf19** [comment]
@ossentoo In the auzre portal, under app services, you will need to make sure App Service Authentication is turned off:
![image](https://user-images.githubusercontent.com/19942418/73690889-29daf980-4686-11ea-8709-b5324996cdd8.png)

**Comment by ossentoo** [comment]
Hi @jennyf19 I've confirmed that Authentication/Authorization is off already.

**Comment by jennyf19** [comment]
@jmprieur thoughts on this?

**Comment by pjevans** [comment]
We're having the exact same issue. Googling came across this post...

**Comment by jennyf19** [comment]
@ossentoo @pjevans Do you have a repro you can share?

**Comment by pjevans** [comment]
I don't, we were trying to deploy front door in front of our core 3.1 app that uses AzureADB2C auth with the supported NuGet package.  Works fantastic until put behind front door, then it just simply doesn't work at all.  I get the same correlation error mentioned above, I looked around and couldn't find anything available to remedy, so we abandoned Front Door.

**Comment by ossentoo** [comment]
@jennyf19 I can try and craft a repo that demos the problem. Will take me a little to do it.

**Comment by jennyf19** [comment]
@ossentoo that would be super helpful. thank you

**Comment by ossentoo** [comment]
@jennyf19 I've just added you to a private fork of this repo: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet

I have amended it to point to an existing deployment of an app service that is deployed behind a front door. Can you point me towards a private channel so we can discuss further?  Although the deployed app service is behind a frontdoor, the configure is currently setup to redirect to azurewebsites.net (just to demo to you that it works, before we make the change). When you are ready, I can make the change so that it redirect back to the frontdoor enabled custom hostname, which will demonstrate the problem I am having.

**Comment by jennyf19** [comment]
@ossentoo thanks so much. you can email me at jeferrie@microsoft.com and include @jmprieur jmprieur@microsoft.com as well? Thank you!

**Comment by Zimmergren** [comment]
I've seen similar issues, and it looks like the redirect uris come back incorrect for my requests. I make a request to the Front Door url, but the b2c returned request persists on containing the azure website backend url as a redirect url (which isn't an allowed redirect uri, as we use the domain for front door). 

Is there any update on this topic, or any output to share from the private e-mail dialogs?

**Comment by EdiWang** [comment]
I have the same problem @Zimmergren 

My website use a custom domain configured on Front Door, but AAD sign in redirect back to the app service default URL (https://****.azurewebsites.net/signin-oidc). 

How to solve this issue? Is there any update?

**Comment by jmprieur** [comment]
@EdiWang @Zimmergren @ossentoo @pjevans 
did you see [Deploy to App Services Linux containers, or with proxys](https://github.com/AzureAD/microsoft-identity-web/wiki/Deploying-Web-apps-to-App-services-as-Linux-containers)?

**Comment by EdiWang** [comment]
I solved the problem by removing the default host header in Azure Front Door backend pool. See my blog post: https://edi.wang/post/2020/6/29/solving-azure-ad-sign-in-failure-with-azure-front-door

**Comment by ossentoo** [comment]
> @EdiWang @Zimmergren @ossentoo @pjevans
> did you see [Deploy to App Services Linux containers, or with proxys](https://github.com/AzureAD/microsoft-identity-web/wiki/Deploying-Web-apps-to-App-services-as-Linux-containers)?

@jmprieur I had previous tried to resolve the issue by manipulating that header.  Anyway, i see that the document has been updated, so I'll take a look.

**Comment by kylepope-ge** [comment]
@ossentoo We are running an ASP.NET Core 3.1 server and have the exact same problem with the correlation cookie. Observing the original request to authenticate, it looks like our server is not including the correct redirect_uri query parameter as mentioned by @Zimmergren. The solution that @EdiWang documented looked promising, but because we don't have a custom domain, going to the Front Door frontend URL results in 404 errors for us.

I've reviewed the link you last referenced and still wasn't able to get it to work either. Setting the `ASPNETCORE_FORWARDEDHEADERS_ENABLED` app config setting to `true` didn't do anything for us for whatever reason and the guide to use the Forwarded Header middleware [here](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1#forwarded-headers-middleware-order) didn't work either.

All that said, I found this solution [here](https://www.phillipsj.net/posts/using-azure-front-door-with-dotnet-core) that worked. It turns out the key is to *also* specify `ForwardedHeaders.XForwardedHost` in the Forwarded Header middleware's configuration. 

Not sure if you figured this out yet or not, but hopefully this helps someone out there...
